mypy/newsemanal/semanal_shared.py

```diff
@@ -190,3 +187,14 @@ def set_callable_name(sig: Type, fdef: FuncDef) -> Type:
             return sig.with_name(fdef.name())
     else:
         return sig
+
+
+def calculate_tuple_fallback(typ: TupleType) -> None:
+    """Calculate a precise item type for the fallback of a tuple type.
+
+    This must be called only after the main semantic analysis pass, since joins
+    aren't available before that.
+    """
```

Comment:
I would mention the chicken and egg problem about tuple fallbacks and checking upper bounds for instances here, and/or maybe in a comment in `semanal_main()` where the two are called.